### DISCUSSION TITLE
PHP errors upon logging a QSO

### DISCUSSION THREAD

**Discussion by msmcguire** [magicbug/Cloudlog#884] [Q&A]
Hi. I'm really enjoying Clouldlog, but recently have been encountering some PHP errors upon logging a QSO. The errors read as follows below. The QSO data is recorded in the log, but each time it throws these errors at the top of the window. Any help is very much appreciated. Thanks.

Mike K9ACM

> A PHP Error was encountered
> Severity: Warning
> 
> Message: Cannot modify header information - headers already sent by (output started at /home/inscape/public_html/k9acm/cloudlog/application/models/Logbook_model.php:2305)
> 
> Filename: controllers/Qso.php
> 
> Line Number: 87
> 
> A PHP Error was encountered
> Severity: Warning
> 
> Message: Cannot modify header information - headers already sent by (output started at /home/inscape/public_html/k9acm/cloudlog/application/models/Logbook_model.php:2305)
> 
> Filename: controllers/Qso.php
> 
> Line Number: 88

**Comment by magicbug** [comment]
Hi Mike

This refers to us setting cookies, are you getting this all the time and are you running any plugins in the browser, certainly haven't changed any part of this code in a couple of years so quite an odd thing.

Other thing is haven't recently changed PHP version or anything like that.

Peter.

**Reply by msmcguire** [reply]
Thanks for the quick reply. It seems that it is happening all the time now. I normally run in Chrome, where I do have a number of plugins. However, I have tested it in Firefox w/out plugins and am getting the same error. Also, I have tested in a couple of mobile browsers on my phone and am getting the same error. It seems to be browser independent. Also, I haven't made any server changes (that I am aware of). I think the error must have started after a an cron job update to Cloudlog, but honestly I have no idea. Any help, of course, is very much appreciated. Thanks.

Mike

**Reply by magicbug** [reply]
Trying to think what might have caused it, does it happen for every callsign logged and can you give an example.

**Reply by msmcguire** [reply]
Well, it's happened with all my recent QSOs and, when I test it with other callsigns, it happens everytime. What do you mean by an example? I'm happy to send whatever info I can to resolve the issue.

**Reply by magicbug** [reply]
Knowing its ALL callsigns is fine.. definitely would like to know what PHP version, also are you running QRZ Logbook with realtime uploads enabled?

**Reply by msmcguire** [reply]
Looks like PHP 7.2.34, Here is the output of php --version:

> PHP 7.2.34 (cli) (built: Feb  4 2021 21:09:06) ( NTS )
> Copyright (c) 1997-2018 The PHP Group
> Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
> with the ionCube PHP Loader + ionCube24 v10.4.5, Copyright (c) 2002-2020, by ionCube Ltd.
> with Zend OPcache v7.2.34, Copyright (c) 1999-2018, by Zend Technologies

Regarding, realtime updates on QRZ Logbook, I didn't see that option before. Perhaps it's relatively new? It was marked as "no." I tried switching it to "yes" but when I went to save the settings, I received another PHP error, pasted below. So might it have something to do with the QRZ integration? Now, I also notice when I go to manually upload the log to QRZ, it says "undefined." This was not the case before.

> A PHP Error was encountered
> Severity: Warning
> 
> Message: Cannot modify header information - headers already sent by (output started at /home/inscape/public_html/k9acm/cloudlog/application/models/Logbook_model.php:2305)
> 
> Filename: helpers/url_helper.php
> 
> Line Number: 564
>

**Reply by msmcguire** [reply]
Any further ideas? I'd love to get this resolve if possible. Thanks again for all your help.

**Reply by msmcguire** [reply]
Ok, so I looked at the post from AndreasK79 below and experimented by adding the following line of PHP code to the index.php file.

`<?php
ob_start(); ?>`

This seems to have corrected the issue. However, I'm not entirely sure what this does or if it might create other problems. Also, I'm not sure if this will be overwritten with automatic updates. Please advise at will. Thanks.

**Reply by AndreasK79** [reply]
@msmcguire seems like you figured out one way to fix it from the link. This will be overwritten with automatic updates. Something is going on with the code, but I haven't tried to figure it out. Perhaps Peter will figure something out after looking at the link I provided.

**Reply by msmcguire** [reply]
@AndreasK79 thanks for the help on this and for looking into it. Do you know if others have been reporting this issue?

**Reply by AndreasK79** [reply]
@msmcguire not that I'm aware of.

**Reply by dl1com** [reply]
> @AndreasK79 thanks for the help on this and for looking into it. Do you know if others have been reporting this issue?

Hi,
I just installed Cloudlog on my server (running PHP 7.3). I also got this warning every time I log a QSO. Also, it was shown after creating my Station Profile.
The fix proposed by @msmcguire worked for me - would be great if we could verify that it has no unwanted side-effects, though :)

**Comment by AndreasK79** [comment]
This might clear some things up:
http://digitalpbk.com/php/warning-cannot-modify-header-information-headers-already-sent

**Reply by msmcguire** [reply]
Thanks for this, but I'm not sure what to do with it. Perhaps Peter can help further with this information?

**Comment by magicbug** [comment]
Just to say that I've made some changes to how cookies are set (Its order) I'm not convinced it will resolve the issue but looking for feedback from people who see this problem

See https://github.com/magicbug/Cloudlog/commit/341c014a8e76b5d631709ae24362cc83706b4f6f
